### PR TITLE
Provide course_type affiliation on courses Sailthru content library

### PR DESCRIPTION
**Pull request by ahsan-ul-haq**
ECOM-6267

Currently it is work in progress and I am working on writing the tests.
@schenedx @tasawernawaz

### CODE REVIEW

File: sailthru_content/services/sailthru_translation_service.py [hunk 1]
```diff
@@ -58,11 +60,23 @@ def _create_course_vars(self, course, course_run, url, site_name, program_dictio
             for seat in course_run['seats']:
                 sailthru_content_vars['price_{}'.format(seat['type'])] = seat['price']
                 sailthru_content_vars['currency_{}'.format(seat['type'])] = seat['currency']
+                if not sailthru_content_vars.get('course_type', None):
+                    sailthru_content_vars['course_type'] = seat['type']
+                else:
```
Comment (schenedx) [inline sailthru_content/services/sailthru_translation_service.py]: Are you thinking that we need to update the existing items in the Sailthru content library this way? We don't need to do that. You can remove this `else` clause

File: sailthru_content/services/sailthru_translation_service.py [hunk 2]
```diff
@@ -58,11 +60,23 @@ def _create_course_vars(self, course, course_run, url, site_name, program_dictio
...
+                    elif sailthru_content_vars['course_type'] == 'verified':
+                        sailthru_content_vars['course_type'] = seat['type']
+                    elif sailthru_content_vars['course_type'] == 'professional':
+                        sailthru_content_vars['course_type'] = seat['type']
                 # add upgrade deadline if there is one
                 if seat.get('upgrade_deadline'):
                     deadline_key = 'upgrade_deadline_{}'.format(seat['type'])
                     sailthru_content_vars[deadline_key] = self._convert_date(seat, 'upgrade_deadline')
 
+        if course_run['key'] in GFA_COURSE_LIST:
```
Comment (schenedx) [inline sailthru_content/services/sailthru_translation_service.py]: What you have in the `GFA_COURSE_LIST` are not course run keys. They are part of the marketing site course about page URL. An example GFA content key is `course-v1:ASUx+ENG102x+2167B`

File: sailthru_content/services/sailthru_translation_service.py [hunk 3]
```diff
@@ -71,6 +85,7 @@ def _create_course_vars(self, course, course_run, url, site_name, program_dictio
                 program = program_dictionary.get(program_link.get('uuid'))
                 if program:
                     sailthru_content_vars['programs'].append(self._translate_program(program))
+                    sailthru_content_vars['course_type'] = program.get('type').lower()
```
Comment (schenedx) [inline sailthru_content/services/sailthru_translation_service.py]: You don't need to send them all lower case. That's not part of the requirement.

**Comment by ahsan-ul-haq** [comment]
@schenedx made some changes please review

File: sailthru_content/services/tests/catalog_api_test_mixins.py [hunk 4]
```diff
@@ -16,7 +16,17 @@ def setUp(self):
         self.api_url_root = 'http://' + faker.domain_name()
         self.catalog_api_service = CatalogApiService(faker.password(), self.oauth_host, None, None, self.api_url_root)
 
-    def prepare_get_courses(self):
+    def prepare_get_courses(self, seat_type=None):
```
Comment (clintonb) [inline sailthru_content/services/tests/catalog_api_test_mixins.py]: As above, import `unicode_literals` and stop using `u`.

File: sailthru_content/services/tests/test_sailthru_translation_service.py [hunk 5]
```diff
@@ -64,3 +67,45 @@ def test_translate_programs(self):
         translated_programs = self.sailthru_translation_service.translate_programs()
         self.assertEqual(len(translated_programs), 1)
         self.assertEqual(len(translated_programs[0]['vars']['course_runs']), 4)
+
+    def _get_expected_sailthru_item(self, seat_type=None):
+        return {
+            'site_name': u'HamiltonX',
```
Comment (clintonb) [inline sailthru_content/services/tests/test_sailthru_translation_service.py]: As above, import `unicode_literals` and stop using `u`.

File: sailthru_content/services/sailthru_translation_service.py [hunk 6]
```diff
@@ -58,11 +60,28 @@ def _create_course_vars(self, course, course_run, url, site_name, program_dictio
             for seat in course_run['seats']:
                 sailthru_content_vars['price_{}'.format(seat['type'])] = seat['price']
                 sailthru_content_vars['currency_{}'.format(seat['type'])] = seat['currency']
+                # Adding the course type with the precedence as
+                # Credit, Professional, Verified, Audit so that if a course run have
+                # multiple seat types the one with the higher precedence will be added
+                # in course_type
+                if not sailthru_content_vars.get('course_type', None):
+                    sailthru_content_vars['course_type'] = seat['type']
+                if sailthru_content_vars['course_type'] == 'audit' or sailthru_content_vars['course_type'] == 'honor':
```
Comment (schenedx) [inline sailthru_content/services/sailthru_translation_service.py]: Why do we need this code block? From line 69 till line 76?

**Comment by schenedx** [comment]
@ahsan-ul-haq I only have one comment. I think this looks pretty good. Can you post the tests you executed? Also, maybe provide a sample set of course_runs objects we are uploading to Sailthru with this change?

File: sailthru_content/services/sailthru_translation_service.py (hunk 6 above)
Comment (ahsan-ul-haq) [inline sailthru_content/services/sailthru_translation_service.py]: A course can have multiple seat types this part of the code make sure that the seat types have precedence as Credit, Professional, Verified, Audit and honor.

Comment (schenedx) [inline sailthru_content/services/sailthru_translation_service.py]: Ah, I didn't notice this code is in a loop on `seats`. We can certainly make this code easier. Something like

```
seat_priority=['professional', 'verified', 'no-id-professional', 'audit', 'honor']
if seat_priority.index(sailthru_content_vars['course_type']) < seat_priority.index(seat['type']):
    sailthru_content_vars['course_type'] = seat['type']
```

File: sailthru_content/services/sailthru_translation_service.py [hunk 7]
```diff
@@ -58,11 +61,23 @@ def _create_course_vars(self, course, course_run, url, site_name, program_dictio
             for seat in course_run['seats']:
                 sailthru_content_vars['price_{}'.format(seat['type'])] = seat['price']
                 sailthru_content_vars['currency_{}'.format(seat['type'])] = seat['currency']
+                # Adding the course type with the precedence as
+                # Credit, Professional, Verified, Audit so that if a course run have
+                # multiple seat types the one with the higher precedence will be added
+                # in course_type
+                if not sailthru_content_vars.get('course_type', None):
+                    sailthru_content_vars['course_type'] = seat['type']
+                if seat_priority.index(sailthru_content_vars['course_type']) < seat_priority.index(seat['type']):
```
Comment (schenedx) [inline sailthru_content/services/sailthru_translation_service.py]: You can combine this if with the if above as:
```
if not sailthru_content_vars.get('course_type', None) or \
    seat_priority.index(sailthru_content_vars['course_type']) < seat_priority.index(seat['type']):
```

File: requirements/base.txt [hunk 8]
```diff
@@ -0,0 +1,9 @@
+edx-rest-api-client==1.6.0
+google-api-python-client==1.5.3
+MySQL-python==1.2.5
+numpy==1.11.1
+pandas==0.16.2
+responses==0.5.1
```
Comment (clintonb) [inline requirements/base.txt]: This is a test package.

File: sailthru_content/services/tests/catalog_api_test_mixins.py [hunk 9]
```diff
@@ -16,7 +17,17 @@ def setUp(self):
...
+        if seat_type:
+            seat = {
+                "type": seat_type,
```
Comment (clintonb) [inline sailthru_content/services/tests/catalog_api_test_mixins.py]: We are using single quotes throughout this file. You should use single quotes here.

File: sailthru_content/services/tests/test_sailthru_translation_service.py [hunk 10]
```diff
@@ -1,12 +1,15 @@
+from __future__ import unicode_literals
 from faker import Factory
```
Comment (clintonb) [inline sailthru_content/services/tests/test_sailthru_translation_service.py]: Remove the newline. All three of these are third-party libraries. There _should_ be a new line after the `__future__` import.

File: sailthru_content/services/sailthru_translation_service.py [hunk 11]
```diff
@@ -58,11 +61,23 @@ def _create_course_vars(self, course, course_run, url, site_name, program_dictio
             for seat in course_run['seats']:
                 sailthru_content_vars['price_{}'.format(seat['type'])] = seat['price']
                 sailthru_content_vars['currency_{}'.format(seat['type'])] = seat['currency']
+                # Adding the course type with the precedence as
```
Comment (clintonb) [inline sailthru_content/services/sailthru_translation_service.py]: Why can't the course run be associated with multiple seat types?

**Pull request merged**